cluster/images/etcd/migrate/migrate_client.go

```diff
@@ -0,0 +1,217 @@
+func (e *CombinedEtcdClient) Snapshot(version *EtcdVersion, snapshotFile string) error {
+	if version.major != 3 {
+		return fmt.Errorf("etcd 3.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"--endpoints", e.endpoint(),
+		"snapshot", "save", snapshotFile,
+	)
+}
+
+// Restore restores a given snapshotFile into the data directory specified this clients config.
+func (e *CombinedEtcdClient) Restore(version *EtcdVersion, snapshotFile string) error {
+	// We cannot use etcd/clientv3 to make this call. It is implemented in the etcdctl client code.
+	if version.major != 3 {
+		return fmt.Errorf("etcd 3.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"snapshot", "restore", snapshotFile,
+		"--data-dir", e.cfg.dataDirectory,
+		"--name", e.cfg.name,
+		"--initial-advertise-peer-urls", e.cfg.peerAdvertiseUrls,
+		"--initial-cluster", e.cfg.initialCluster,
+	)
+}
+
+// Migrate upgrades a 'etcd2' storage version data directory to a 'etcd3' storage version
+// data directory.
+func (e *CombinedEtcdClient) Migrate(version *EtcdVersion) error {
+	// We cannot use etcd/clientv3 to make this call as it is implemented in etcd/etcdctl.
+	if version.major != 3 {
+		return fmt.Errorf("etcd 3.x required but got version '%s'", version)
+	}
+	return e.runEtcdctlCommand(version,
+		"migrate",
+		fmt.Sprintf("--data-dir=%s", e.cfg.dataDirectory),
+	)
+}
+
+func (e *CombinedEtcdClient) runEtcdctlCommand(version *EtcdVersion, args ...string) error {
+	etcdctlCmd := exec.Command(filepath.Join(e.cfg.binPath, fmt.Sprintf("etcdctl-%s", version)), args...)
+	etcdctlCmd.Env = []string{fmt.Sprintf("ETCDCTL_API=%d", version.major)}
+	etcdctlCmd.Stdout = os.Stdout
+	etcdctlCmd.Stderr = os.Stderr
+	return etcdctlCmd.Run()
+}
+
+// AttachLease attaches leases of the given leaseDuration to all the  etcd objects under
+// ttlKeysDirectory specified in this client's config.
+func (e *CombinedEtcdClient) AttachLease(leaseDuration time.Duration) error {
+	ttlKeysPrefix := e.cfg.ttlKeysDirectory
+	// Make sure that ttlKeysPrefix is ended with "/" so that we only get children "directories".
+	if !strings.HasSuffix(ttlKeysPrefix, "/") {
+		ttlKeysPrefix += "/"
+	}
+	ctx := context.Background()
+
+	v3client, err := e.clientV3()
+	if err != nil {
+		return err
+	}
+	defer v3client.Close()
+	objectsResp, err := v3client.KV.Get(ctx, ttlKeysPrefix, clientv3.WithPrefix())
+	if err != nil {
+		return fmt.Errorf("Error while getting objects to attach to the lease")
+	}
+
+	lease, err := v3client.Lease.Grant(ctx, int64(leaseDuration/time.Second))
+	if err != nil {
+		return fmt.Errorf("Error while creating lease: %v", err)
+	}
+	glog.Infof("Lease with TTL: %v created", lease.TTL)
+
+	glog.Infof("Attaching lease to %d entries", len(objectsResp.Kvs))
+	for _, kv := range objectsResp.Kvs {
+		_, err := v3client.KV.Put(ctx, string(kv.Key), string(kv.Value), clientv3.WithLease(lease.ID))
```

Comment:
If the etcd instance if recent enough to support PrevKv, would it be worth validating that it matches the put key/value?

Reply:
Good question. This is the code we moved over from attachlease.go.  I'll go ahead and add the check-- if it fails that strongly suggests concurrent access to the etcd instance while the migration is being run, which would be very bad.